File: pinocchio_robotic_system.py
import numpy as np
import pinocchio as pin
import yaml
from mechanical_system import MechanicalSystem  # Import the base class

class PinocchioRoboticSystem(MechanicalSystem):
    def __init__(self, st, model_name):
        """
        Initialize the Pinocchio-based robotic system with a given sampling period and model.
        """
        super().__init__(st)  # Call parent constructor
        
        # Load model from URDF file
        self.model = pin.buildModelFromUrdf(f'+{model_name}/model.urdf')
        self.data = self.model.createData()
        
        # Load simulation parameters from YAML file
        with open(f'+{model_name}/config.yaml', 'r') as file:
            config = yaml.safe_load(file)
        
        # Extract parameters from YAML configuration
        self.K = np.array(config['K'])  # Joint stiffness matrix
        self.D = np.array(config['D'])  # Joint damping matrix
        self.motor_inertia = np.diag(np.array(config['motor_inertia']))  # Motor inertia
        self.motor_viscous_term = np.array(config['motor_viscous_term'])  # Motor viscous friction
        self.motor_coulomb_term = np.array(config['motor_coulomb_term'])  # Motor Coulomb friction
        self.umax = np.array(config['umax'])  # Maximum control input
        self.x0 = np.array(config['x0'])  # Initial state
        noise = np.array(config['noise'])  # Measurement noise
        self.u = np.array(config['u0'])  # Control input initialization
        
        # Define joint and control-related properties
        self.njoints = int(self.model.nq)  # Number of joints
        self.n_controlled_joints = self.njoints  # Number of controlled joints
        self.num_input = self.n_controlled_joints
        self.num_output = 2 * self.n_controlled_joints
        self.order = self.njoints * 2  # System order
        
        # Define transformation matrices
        self.stateToOutput = np.block([
            [np.zeros((self.njoints, self.njoints)), np.eye(self.njoints), np.zeros((self.njoints, 2 * self.njoints))],
            [np.zeros((self.njoints, 3 * self.njoints)), np.eye(self.njoints)]
        ])
        
        self.u0 = np.zeros(self.n_controlled_joints)  # Initial control action
        self.u = self.u0.copy()
        
        # Define noise levels for sensor measurements
        self.sigma_y = np.concatenate([
            noise[0] * np.ones(self.n_controlled_joints),
            noise[1] * np.ones(self.n_controlled_joints)
        ])
        
        # Initialize state
        self.x = self.x0.copy()
        
        # Define output and input names
        self.output_names = [f'position_{i+1}' for i in range(self.n_controlled_joints)] + \
                            [f'velocity_{i+1}' for i in range(self.n_controlled_joints)]
        self.input_names = [f'torque_{i+1}' for i in range(self.n_controlled_joints)]
        
        self.payload = 0  # Initialize payload
    
    def full_joint_position(self):
        """
        Return the full joint position vector.
        """
        return self.x[:self.njoints]
    
    def set_payload(self, payload):
        """
        Set the payload value.
        """
        self.payload = payload
    
    def state_function(self, x, u, t):
        """
        Compute the state derivative dx/dt given state x and input u.
        """
        ql = x[:self.njoints]  # Link position
        qm = x[self.njoints:2*self.njoints]  # Motor position
        qld = x[2*self.njoints:3*self.njoints]  # Link velocity
        qmd = x[3*self.njoints:4*self.njoints]  # Motor velocity
        
        # Compute elastic torque
        tau_elastic = self.K * (qm - ql) + self.D * (qmd - qld)
        
        # Apply saturation to control input
        u = np.clip(u, -self.umax, self.umax)
        
        # Convert state to numpy arrays for Pinocchio
        q_numpy = np.array(ql)
        qp_numpy = np.array(qld)
        qpp_numpy = np.zeros_like(ql)
        
        # Compute nonlinear torque using inverse dynamics
        nonlinear_tau = pin.rnea(self.model, self.data, q_numpy, qp_numpy, qpp_numpy)
        inertia = pin.crba(self.model, self.data, q_numpy)
        
        # Solve for link acceleration
        qldd = np.linalg.solve(inertia, tau_elastic - nonlinear_tau)
        
        # Solve for motor acceleration
        qmdd = np.linalg.solve(
            self.motor_inertia, u - tau_elastic - self.motor_viscous_term * qmd - self.motor_coulomb_term * np.tanh(30 * qmd)
        )
        
        return np.concatenate([qld, qmd, qldd, qmdd])
    
    def output_function(self):
        """
        Compute the output of the system given the current state.
        """
        return self.stateToOutput @ self.x + self.sigma_y * np.random.randn(len(self.sigma_y))

File: PinocchioRoboticSystem.m
classdef PinocchioRoboticSystem < MechanicalSystem
    properties  (Access = protected)
        njoints
        n_controlled_joints
        controlled_joints
        inputToState
        stateToOutput
        payload
        tool_mass
        model
        data
        K
        D
        motor_inertia
        motor_viscous_term
        motor_coulomb_term
    end

    methods  (Access = public)
        function obj=PinocchioRoboticSystem(st,model_name)
            obj@MechanicalSystem(st); % uso il costruttore parente
            obj.model=py.pinocchio.buildModelFromUrdf(sprintf('+%s/model.urdf',model_name));

            obj.data=obj.model.createData();
            
            
            [obj.K,obj.D,obj.motor_inertia,obj.motor_viscous_term,obj.motor_coulomb_term,obj.umax,obj.x0,noise,obj.u]=eval(sprintf('%s.dataSimulation',model_name));

            n=double(obj.model.nq);
            obj.njoints=n;
            obj.stateToOutput=[zeros(n) eye(n) zeros(n,2*n);
                               zeros(n,3*n) eye(n)];
            
            obj.n_controlled_joints=obj.njoints;
            obj.num_input=obj.n_controlled_joints;
            obj.num_output=2*obj.n_controlled_joints;

            obj.order=obj.njoints*2;
            obj.u0=zeros(obj.n_controlled_joints,1);
            obj.u=obj.u0;
            

            obj.sigma_y=[noise(1)*ones(obj.n_controlled_joints,1); noise(2)*ones(obj.n_controlled_joints,1)];
            
            % initial inital control action to stay close to steady state.
            obj.x=obj.x0;
            
            for idx=1:obj.n_controlled_joints
                obj.output_names{idx}=sprintf('position_%d',idx);
                obj.output_names{idx+obj.n_controlled_joints}=sprintf('velocity_%d',idx);
            end

            for idx=1:obj.n_controlled_joints
                obj.input_names{idx}=sprintf('torque_%d',idx);
            end

            obj.payload=0;
        end

        function full_y=fullJointPosition(obj)
            full_y=obj.x(1:obj.njoints);
        end

        function setPayload(obj,payload)
            obj.payload=payload;
        end
    end



    methods  (Access = protected)
        function Dx=stateFunction(obj,x,u,t)

            ql = x((1:obj.njoints));  % pos link
            qm = x((1:obj.njoints)+obj.njoints); %pos motor
            qld= x((1:obj.njoints)+2*obj.njoints); %vel link
            qmd= x((1:obj.njoints)+3*obj.njoints); %vel link

            tau_elastic=obj.K*(qm-ql)+obj.D*(qmd-qld);

            u=max(min(obj.umax,u),-obj.umax);
            
            q_numpy=py.numpy.array(ql);
            qp_numpy=py.numpy.array(qld);
            qpp_numpy=py.numpy.array(0*ql);
            nonlinear_tau = double(py.pinocchio.rnea(obj.model, obj.data, q_numpy,qp_numpy,qpp_numpy))';
            inertia=double(py.pinocchio.crba(obj.model,obj.data,q_numpy));
            
            %tau = inertia*qldd + nonlinear(ql,qld) 
            qldd = inertia\(tau_elastic-nonlinear_tau);

            qmdd=obj.motor_inertia\(u-tau_elastic-obj.motor_viscous_term*qmd-obj.motor_coulomb_term*tanh(30*qmd));
            Dx=[qld;qmd;qldd;qmdd];
        end
        function y=outputFunction(obj)
            y=obj.stateToOutput*obj.x+obj.sigma_y.*randn(length(obj.sigma_y),1);
        end

    end
end